streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -170,28 +167,31 @@ public boolean initializeStateStores() {
     @Override
     public void initializeTopology() {
         initTopology();
+
+        // if (eos) begin new transaction
+        if (eosEnabled) {
+            try {
+                this.producer.beginTransaction();
+            } catch (final ProducerFencedException fatal) {
+                throw new TaskMigratedException(this, fatal);
+            }
+            transactionInFlight = true;
+        }
+
         processorContext.initialized();
         taskInitialized = true;
     }
 
     /**
      * <pre>
      * - re-initialize the task
-     * - if (eos) begin new transaction
      * </pre>
      * @throws TaskMigratedException if the task producer got fenced (EOS only)
```

Comment:
This cannot happen anymore and we should remove this line

Reply:
That makes sense. 

Have also updated the calling path's throw comments to reflect this change.